libqtile/widget/currentlayout.py

```diff
@@ -57,3 +69,118 @@ def button_press(self, x, y, button):
             self.qtile.cmd_next_layout()
         elif button == 2:
             self.qtile.cmd_prev_layout()
+
+
+class CurrentLayoutIcon(base._TextBox):
+    """
+    Display the icon representing the current layout of the
+    current group of the screen, the bar containing the widget, is on.
+    """
+    orientations = base.ORIENTATION_HORIZONTAL
+
+    defaults = [
+        ('scale', 1, 'Scale factor, defaults to 1'),
+        ('theme_path', default_icon_path(), 'Path to layout icons')
+    ]
+
+    def __init__(self, **config):
+        base._TextBox.__init__(self, "", **config)
+        self.add_defaults(CurrentLayoutIcon.defaults)
+        self.scale = 1.0 / self.scale
+
+        self.length_type = bar.STATIC
+        self.length = 0
+
+    def _configure(self, qtile, bar):
+        base._TextBox._configure(self, qtile, bar)
+        self.text = self.bar.screen.group.layouts[0].name
+        self.icons_loaded = False
+        self.surfaces = {}
+        self.setup_images()
+        self.setup_hooks()
+
+    def setup_hooks(self):
+        """
+        Listens for layout change and performs a redraw when it occurs.
+        """
+        def hook_response(layout, group):
+            if group.screen is not None and group.screen == self.bar.screen:
+                self.current_layout = layout.name
+                self.bar.draw()
+        hook.subscribe.layout_change(hook_response)
+
+    def button_press(self, x, y, button):
+        if button == 1:
+            self.qtile.cmd_next_layout()
+        elif button == 2:
+            self.qtile.cmd_prev_layout()
+
+    def draw(self):
+        if self.icons_loaded:
+            try:
+                surface = self.surfaces[self.current_layout]
+            except KeyError:
+                logger.error('No icon for layout {}'.format(
+                    self.current_layout
+                ))
+            else:
+                self.drawer.clear(self.background or self.bar.background)
+                self.drawer.ctx.set_source(surface)
+                self.drawer.ctx.paint()
+                self.drawer.draw(offsetx=self.offset, width=self.length)
+        else:
+            # Fallback to text
+            self.text = self.current_layout[0].upper()
+            base._TextBox.draw(self)
+
+    def _get_layout_names(self):
+        """
+        Returns the list of lowercased strings for each available layout name.
+        """
+        return [
+            layout_class_name.lower()
+            for layout_class, layout_class_name
+            in map(lambda x: (getattr(layout_module, x), x), dir(layout_module))
+            if isinstance(layout_class, (type, types.ClassType)) and issubclass(layout_class, Layout)
+        ]
+
+    def setup_images(self):
+        """
+        Loads layout icons.
+        """
+        for key in self._get_layout_names():
```

Comment:
My fear with this is it becomes a bit more difficult to add new layouts to this repo without adding a new icon, and on the flip side, when users create custom layouts in their config, they won't be picked up in _get_layout_names. I'm not sure how big of a problem either of these things are and would be up to suggestions on how to fix them (if they are actually problematic).